src/twisted/internet/asyncfs.py

```diff
@@ -0,0 +1,709 @@
+                self._buffer = bytearray()
+                if len(buf) == 0:
+                    self.event.clear()
+            if len(buf) == 0:
+                if self._closed:
+                    break
+                if self.producer and (
+                    (self.streaming and self._paused) or (not self.streaming)
+                ):
+                    self._paused = False
+                    blockingCallFromThread(
+                        reactor, cast("IPushProducer", self.producer).resumeProducing
+                    )
+                self.event.wait()
+            else:
+                os.write(self.fd, buf)
+
+
+@implementer(IAsyncFilesystem)
+class ThreadFs:
+    """
+    An async filesystem implemented using threads
+
+    This implementation cribbed from
+    L{twisted.conch.unix.SFTPServerForUnixConchUser}
+
+    However unlike conch we don't have per-user processes so more of our own security.
+    """
+
+    def __init__(
+        self,
+        root: str,
+        threadpool: Optional["ThreadPool"] = None,
+        read_only: bool = False,
+    ) -> None:
+        """
+        @param root: the base directory
+        @type root: L{str}
+
+        @param threadpool: the threadpool to use, default the reactor's threadpool
+        @type threadpool: L{twisted.python.threadpool.ThreadPool}
+
+        @param read_only: if C{True}, changes to filesystem forbidden
+        @type read_only: C{bool}
+        """
+        self.root = root
+        self.read_only = read_only
+        if threadpool:
+            self.threadpool = threadpool
+        else:
+            self.threadpool = reactor.getThreadPool()  # type: ignore
+
+    def unregister(self) -> None:
+        pass
+
+    def _setAttrs(self, path: str, attrs: dict) -> bool:
+        if "uid" in attrs and "gid" in attrs:
+            os.chown(path, attrs["uid"], attrs["gid"])
+        if "permissions" in attrs:
+            os.chmod(path, attrs["permissions"])
+        if "atime" in attrs and "mtime" in attrs:
+            os.utime(path, (attrs["atime"], attrs["mtime"]))
+        return True
+
+    def _getAttrs(self, s: os.stat_result) -> dict:
+        d = {
+            "size": s.st_size,
+            "uid": s.st_uid,
+            "gid": s.st_gid,
+            "permissions": s.st_mode,
+            "atime": int(s.st_atime),
+            "mtime": int(s.st_mtime),
+            "ext_ctime": int(s.st_ctime),
+            "ext_nlinks": s.st_nlink,
+        }
+        try:
+            d["ext_blksize"] = s.st_blksize
+        except AttributeError:
+            # not available on all platforms
+            pass
+        try:
+            if s.st_birthtime:
+                d["ext_birthtime"] = s.st_birthtime
+        except AttributeError:
+            # not available on all platforms
+            pass
+        return d
+
+    def _absPath(self, path: str) -> str:
+        p = os.path.normpath(os.path.join(self.root, path))
+        if not p.startswith(self.root):
```

Comment:
What if `self.root == "/foo"` and `p == "/foobar"`?  It would be good to use FilePath everywhere to avoid this kind of problem.